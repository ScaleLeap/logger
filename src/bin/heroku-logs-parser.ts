#!/usr/bin/env node

import { HerokuLogsParser } from '../heroku-logs-parser'

const hp = new HerokuLogsParser()

process.stdin.pipe(hp).pipe(process.stdout)
