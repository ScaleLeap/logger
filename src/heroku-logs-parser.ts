#!/usr/bin/env node

import { Transform, TransformCallback } from 'stream'

const PREFIX = new RegExp(
  /^(\d{4}-[01]\d-[0-3]\dT[0-2]\d:[0-5]\d:[0-6]\d(\.\d{3,6})?\+\d\d:\d\d\s\w+\[.+]:\s?)/,
)

export function parseLines(lines: string[]) {
  return lines.map((line) => line.replace(PREFIX, ''))
}

export class HerokuLogsParser extends Transform {
  _transform(chunk: Buffer, encoding: string, callback: TransformCallback) {
    this.push(parseLines(chunk.toString().split('\n')).join('\n'))

    callback()
  }
}
