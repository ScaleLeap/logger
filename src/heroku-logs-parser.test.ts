import { HerokuLogsParser, parseLines } from './heroku-logs-parser'

const lines = `
2020-05-05T09:13:34.754509+00:00 app[run.1787]: {"level":30,"time":1588670014754,"pid":115,"hostname":"a8d57749-9c74-4a85-8edc-x","msg":"msg1"}
2020-05-05T09:13:34.759052+00:00 app[run.1787]: {"level":30,"time":1588670014757,"pid":115,"hostname":"a8d57749-9c74-4a85-8edc-x","msg":"msg2"}
2020-05-05T09:12:48+00:00 app[heroku-postgres]: source=HEROKU_POSTGRESQL addon=postgresql-animate-13221 sample#current_transaction=2836
2020-05-05T09:14:44.550886+00:00 app[run.1787]: {"level":30,"time":1588670084548,"pid":115,"hostname":"a8d57749-9c74-4a85-8edc-x","msg":"msg3"}
`
  .trim()
  .split('\n')

describe(`${HerokuLogsParser.name}`, () => {
  it('should parse lines', () => {
    expect.assertions(1)

    expect(parseLines(lines)).toMatchSnapshot()
  })
})
