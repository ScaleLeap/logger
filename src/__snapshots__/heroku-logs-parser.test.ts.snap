// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HerokuLogsParser should parse lines 1`] = `
Array [
  "{\\"level\\":30,\\"time\\":1588670014754,\\"pid\\":115,\\"hostname\\":\\"a8d57749-9c74-4a85-8edc-x\\",\\"msg\\":\\"msg1\\"}",
  "{\\"level\\":30,\\"time\\":1588670014757,\\"pid\\":115,\\"hostname\\":\\"a8d57749-9c74-4a85-8edc-x\\",\\"msg\\":\\"msg2\\"}",
  "source=HEROKU_POSTGRESQL addon=postgresql-animate-13221 sample#current_transaction=2836",
  "{\\"level\\":30,\\"time\\":1588670084548,\\"pid\\":115,\\"hostname\\":\\"a8d57749-9c74-4a85-8edc-x\\",\\"msg\\":\\"msg3\\"}",
]
`;
